### ISSUE TITLE
messageRateMax - I think we might need a separate post_messageRateMax

### ISSUE THREAD

**Issue by reidsunderland** [MetPX/sarracenia#1181]
I tried using messageRateMax to slow down publishing in a poll, but it appears to also slow down ingestion of messages into the nodupe cache.

In a poll, I think we'd mostly want to only limit the rate messages are posted at, not ingested. So maybe we need `post_messageRateMax` and make the "plain" `messageRateMax` only for incoming messages?

**Comment by petersilva** [comment]
@mshak2 probably your next target... ;-)

**Comment by petersilva** [comment]
same as #1213

**Comment by reidsunderland** [comment]
This came up again while @andreleblanc11 was working on some polls.

We talked about a potential workaround using a shovel:

poll --> shovel with messageRateMax --> sarras

That should deliver a rate limited stream of messages to the sarras.

**Comment by petersilva** [comment]
so... questions come up.  Since this means actively queueing in the application (sr3) rather than using the broker.  The most obvious method is the retry queue, so whenever you exceed the publication rate, you take the extra messages and put them in the retry_queue.

They will get retried after 5 minutes (by default) and newer messages in the mean-time will get through.  Is that what you want/had in mind?

**Comment by reidsunderland** [comment]
Yeah that's a good point. I don't think I had any potential implementations in mind when I created the issue, I just saw a problem.

If each poll run returns 1000 messages, but we only want to post a few of them (by accept/reject statements), then the rate limit could slow the poll down to an almost unusable rate.

In the case we're dealing with, we have to respect a rate limit of < 120 requests per minute. I think the rate limiting in the sarra wasn't working well enough across the 8 nodes in the cluster (or maybe I thought that the poll itself was also exceeding the rate limit).

I agree that using the broker to queue messages would be better than queueing inside sr3. Having messages disappear and come back after 5 minutes would probably be confusing for users. And the disk queue adds complexity when it comes to VIP switches. 

Maybe closing this and using the shovel workaround as the preferred solution would be good enough?

**Comment by petersilva** [comment]
Hang on, if you have eight nodes polling in sr3... only one of them should be active.
The others should just be updating their post queues based on what the active one did.
It is supposed to be much better than v2 for that (not 8 nodes all polling.) 

## Alternate resolution:

I'm wondering if we can do something similar on the polling side, that is, if the rate is exceeded, we drop files we see before they get into the duplicate suppression cache, so that at the next polling interval, it still shows up as new.  That should be pretty easy...   and use the same messageRateMax to control it.  Is that helpful?  A further step would be to just stop polling entirely until the messageRateMax allows it.  also likely easy.

This way, we are not queueing in the app, and not creating any extra load for anyone (all the candidates show up in the poll anyways, we're just changing when we mark some to be ignored.)

**Comment by reidsunderland** [comment]
Yes, just one node polls, but the 8 nodes download. It's really the download where we hit the rate limit of the source. It's been a long time now, so I can't remember why using the rate limiting in the sarra wasn't an option. I may have been trying to do the rate limiting in the poll so I could use an sr3 poll to feed a v2 sarra (with no rate limiting).

The poll itself shouldn't really be making many requests to the server, only ~1 request per directory being polled?

I think either of the proposed options might work, but the more I think about it the less I understand why we wouldn't just rate limit in the sarra.

**Comment by petersilva** [comment]
OK, I was thinking about how to rate limit in the poll, looked over the code, and realized the poll should already do that. Looking at the first post in the issue, I realized that you tried that, and it does work, but by slowing down how quickly the poll scans the source... I'm thinking that's probably right and good as-is.  

What is the problem with just using messageRateMax in the poll as-is?

**Comment by reidsunderland** [comment]
Yes I think you're caught up :)

The way it's currently implemented, the rate limiting uses the `last_gather_len` value, which is the number of messages gathered **before filtering**.  In a poll, usually `last_gather_len` is significantly larger than the number of messages that eventually get posted (after accept/reject filtering and duplicate suppression).

So yes, the current rate limiting implementation does slow down how quickly we poll the source.

When I created this issue, I only wanted to limit the poll's publish rate (i.e. apply the rate limit to the much smaller number of messages that remain post-filtering). The polling itself was already slow enough. 

After thinking about it more, I'm 99% sure the reason I wanted that feature was to allow me to feed messages to v2 sarras slowly. 

Moving the sarras to sr3 and using the rate limit on the sarra is a much better solution, so I'm okay with giving up on this and closing it.

**Comment by petersilva** [comment]
Yeah, it is quite normal that poll and watch can see huge numbers of rejects in the filtering.  Two questions arise:

1. maybe we're doing it wrong: we should limit all components based on the post-filter rate.

2. lesser change: maybe we should count only after filter for poll and watch?

Either of these might be a good idea.

**Comment by reidsunderland** [comment]
That's a good point. I actually had a problem with rate limiting on a sender a few months ago.

On DDSR dev, we have 1 queue per node and have `accept .*${HOSTNAME}...` in the config. So if I want to send at a rate of 1 message per second, I actually have to set the `messageRateMax` to `2` (because in that kind of 2 node cluster with no local brokers, ~50% of messages are always rejected). 

So I do think a post-filter rate limit would be useful.

**Comment by reidsunderland** [comment]
I think I'm okay with just option 1, change the messageRateMax to be post-filtering

If pre-filtering rate limiting was needed, it would be easy enough to re-implement in the main code or maybe in an after_gather plugin

**Comment by petersilva** [comment]
Looking at the code... last_gather_len is  used for messageCountMax also.   Thinking about it, I think post filtering total makes more sense there also?
What do you think?

**Comment by reidsunderland** [comment]
Yeah, I agree that the post-filter count makes sense for messageCountMax.
